newIDE/app/public/external/piskel/piskel-main.js

```diff
@@ -281,87 +153,204 @@ function piskelCreateAnimation(pskl, piskelOptions) {
-  const imageData = [];
-  let maxWidth = -1;
-  let maxHeight = -1;
-  async.each(
-    piskelOptions.resources,
-    function(resource, callback) {
-      const image = new Image();
-      image.onload = function() {
-        imageData.push(image);
-        maxWidth = Math.max(image.width, maxWidth);
-        maxHeight = Math.max(image.height, maxHeight);
-        callback();
-      };
 
-      // onload will fire after `src` is set
-      try {
-        image.src = readBase64ImageFile(resource.resourcePath);
-      } catch (error) {
-        // Unable to load the image, ignore it.
-        console.error('Unable to load ', resource, error);
-        callback();
+  // If there is metadata, use it to load the frames with layers
+  // Note that metadata will be saved only if the user has more than one layers
+  if (piskelOptions.externalEditorData.pskl) {
+    const editorDataPaths = piskelOptions.externalEditorData.pskl.paths;
+
+    // Create a Piskel Document from the metadata that GD stores
+    pskl.utils.serialization.Deserializer.deserialize(
+      JSON.parse(piskelOptions.externalEditorData.pskl.data),
+      piskel => {
+        piskelController.setPiskel(piskel);
+        // set piskel frame paths to their piskel data counterpart - on all layers
+        for (let i = 0; i < piskelController.getFrameCount(); i++) {
+          piskelController.getLayers().forEach(layer => {
+            layer.getFrameAt(i).originalPath = editorDataPaths[i];
+            layer.getFrameAt(i).originalIndex = i;
+          });
+        }
+
+        // Compare the imported frames - so as to make the layered Piskel Document
+        // the same as the changes done in Gdevelop without flattening any layers
+        let flattenedImagePaths = [];
+        piskelOptions.resources.forEach((resource, frameIndex) => {
+          const flattenedFramePath = path.normalize(resource.resourcePath);
+          flattenedImagePaths.push(flattenedFramePath);
+
+          // Import any frames that were added in Gdevelop
+          if (!editorDataPaths.includes(flattenedFramePath)) {
+            pskl.utils.BlobUtils.dataToBlob(
+              readBase64ImageFile(flattenedFramePath),
+              'image/png',
+              imageBlob => {
+                pskl.utils.FileUtils.readImageFile(imageBlob, image => {
+                  piskelController.setCurrentFrameIndex(
+                    piskelController.getFrameCount() - 1
+                  );
+
+                  piskelController.addFrameAtCurrentIndex();
+                  piskelController.selectNextFrame();
+                  const currentFrameObj = piskelController.getCurrentFrame();
+                  pskl.utils.FrameUtils.addImageToFrame(
+                    currentFrameObj,
+                    image,
+                    0,
+                    0
+                  );
+                  pskl.tools.transform.TransformUtils.center(currentFrameObj);
+
+                  piskelController.getLayers().forEach(layer => {
```

Comment:
Ideally, you could extract all the "layer and frames manipulation" into tool functions :)